### PR TITLE
Stale cache

### PR DESCRIPTION
**Pull request by joehoyle**
## Allow serving of stale cache

Typically when chosing a TTL for cache, it's a trade off between performance for the end user and how long you want to be serving out of date content on your page (purgin aside). For high traffic this often isn't an issue, as the percentage of users hitting uncached content is small. 100 visits a minute with a cache TTL of 5 minutes will only result in on in 500 visitors experience a (relatively) slow load time.

The problem with this approach is for low traffic sites. If you have one visitor every hour, it's likely those visitors will always be hitting an uncached page and experiance a slow site. There is not enough visitors to keep the cache "warm". This means the percentage of visitors having a bad experience is much higher. One option would be to increase cache TTL number to something very large, like 2 hours - however this creates likely hood of serving _very_ stale content, and in the event of a traffic spike, one would have benefitted from using a shorter TTL and be able to serve updated content in a more timely manor.

This pull-request aims to make this process a little more robust by passing a "stale" cache to the user loading the site, and then building the page (with any updated content) and pushing it to the cache for the next visit. Varnish calls this a `grace` response, serve an expired item while simultaniously sending a request to the backend to update the cache for the next person.

Nginx doesn't yet offer this functionality, though there is a Bounty to add support: https://www.bountysource.com/issues/972735-proxy_cache_use_stale-run-updating-in-new-thread-and-serve-stale-data-to-all

With this pull request, let's assume `max_age` is set to 15 seconds, and stale_max_age is set to 120 seconds. This will happen from non cache existing:
1. 0 Seconds - User visits, cache `MISS`, user waits for page to be generated
2. 10 Seconds - User visits, cache `HIT`, user is given a page generated at `0 seconds`
3. 40 seconds - User visits, cache `UPDATING`, user is given a page genereted at `0 seconds`
4. 50 seconds - User visits, cache `HIT`, user is given a page genreated at `40 seconds`

With the above flow, the user is given a page within circa 30 milliseconds while still getting the latest content. This is all made possible with the PHP-FPM function `fastcgi_finish_request()`
 which is used to push the stale cache item to the user, close the connection and continue generating the page as if it were a `MISS` to repopualte the cache with the updated version of the pag
e.

There are a couple of kinks to work out still, as this is somewhat voodoo.
1. It's not currently possible to set new headers if a stale page has been served, as PHP Core doesn't have much of a concept or consideration for `fastcgi_finish_request()` if you have output your content already, it won't allow the use of `header()`, so Batcache's uses of `sent_headers()` doesn't function.
2. Stale cache serving for redirects / IF-Modified-Since is not currently supported

### CODE REVIEW

**Comment by Viper007Bond** [comment]
Batcache was originally introduced to reduce server load rather than speed up page requests I believe so this would be an interesting change. Of course current functionality could be to set both values to the same so that there is no stale cache serving.

Anyway, Batcache isn't really my baby, I just took it upon myself to stick the latest internal version here on GitHub. So pinging @mdawaffe who can probably weigh in here better than I.

File: advanced-cache.php [hunk 1]
```diff
@@ -208,7 +237,19 @@ function ob($output) {
 
 		$this->cache['max_age'] = $this->max_age;
 
-		wp_cache_set($this->key, $this->cache, $this->group, $this->max_age + $this->seconds + 30);
+		if ( $this->use_stale ) {
+
+			$expire = $this->stale_max_age + $this->max_age + $this->seconds + 30;
```
Comment (mdawaffe) [inline advanced-cache.php]: I would put this in an `else` of the following `if` statement.

File: advanced-cache.php [hunk 2]
```diff
@@ -319,34 +365,68 @@ function add_debug_html_to_output( $debug_html ) {
...
+	if ( $batcache->add_hit_status_header ) {
+		header( 'X-Batcache: BYPASS' );
+	}
+
 	return;
+}
 
 // Never batcache when POST data is present.
-if ( ! empty( $GLOBALS['HTTP_RAW_POST_DATA'] ) || ! empty( $_POST ) )
+if ( ! empty( $GLOBALS['HTTP_RAW_POST_DATA'] ) || ! empty( $_POST ) || $_SERVER['REQUEST_METHOD'] === "POST" ) {
```
Comment (mdawaffe) [inline advanced-cache.php]: This looks like a separate issue, so should go in a separate PR.

File: advanced-cache.php [hunk 3]
```diff
@@ -432,7 +512,7 @@ function add_debug_html_to_output( $debug_html ) {
 
 
 // Did we find a batcached page that hasn't expired?
-if ( isset($batcache->cache['time']) && ! $batcache->genlock && time() < $batcache->cache['time'] + $batcache->cache['max_age'] ) {
+if ( isset($batcache->cache['time']) && ! $batcache->genlock && ( time() < $batcache->cache['time'] + $batcache->cache['max_age'] || ( $batcache->use_stale && time() < $batcache->cache['time'] + $batcache->cache['max_age'] + $batcache->stale_max_age ) || ( $batcache->use_stale && $batcache->stale_max_age === 0 ) ) ) {
```
Comment (mdawaffe) [inline advanced-cache.php]: Parenthesis inconsistency: the expired check is not wrapped in parentheses, but the use_stale checks both are.

(The inconsistency is just a style issue: it doesn't cause any functional issue.)

Comment (mdawaffe) [inline advanced-cache.php]: I think there's a bug where if use_stale is set, but `fastcgi_finish_request()` doesn't exist, this patch will always respond with cached content, and the cache will never get cleared/updated.

**Comment by mdawaffe** [comment]
Idea seems sane to me.

@skeltoac is the source of truth here, though :)

**Comment by Viper007Bond** [comment]
Oh right, right, sorry @mdawaffe for the wrong ping. Not sure why I was thinking this was your thing.

**Comment by skeltoac** [comment]
Interesting. The new piece is an option to serve stale content in favor of waiting the page generation time, potentially speeding up one page view per TTL.

It looks like this was added without due regard for stampedes. There is already `genlock` in batcache which is stampede protection. Notice that we never `wp_cache_delete()` the cache itself.

The idea is a definite improvement and I am glad you are planning on contributing.

**Comment by toddlahman** [comment]
In regards to `genlock`.

https://github.com/Automattic/batcache/issues/34

**Comment by rmccue** [comment]
> It's not currently possible to set new headers if a stale page has been served

Thinking at the moment, we're better off to send a loopback request to ourselves to ensure this isn't a problem. Got the code ready for that. :)

**Comment by vnsavage** [comment]
A use_stale parameter was added to facilitate serving of stale responses during the generation of new cache. It differs from the Varnish 'grace' behavior in that it does not return stale responses, on backend/generation errors.

The idea of returning stale responses if they're available sounds good in general, but would need a new PR as:
1) It shouldn't rely on fastcgi_finish_request() being available.
2) It looks like executing wp_cache_delete() can cause race conditions when Batcache is used in multi-datacenter environment with cache replication.
3) There have been too many changes after this PR was submitted, so it will need to be re-created.
4) The TODOs in the initial request and from the code reviews here need to be implemented.

**Comment by joehoyle** [comment]
I haven't picked this up in a while - I hit a bit of a wall with trying to solve the cookie issue (when trying to serve stale _and_ regenerate). I think @rmccue's suggestion of a loopback would be in all cases, reducing the reliance on `fastcgi_finish_request`.

I'de rather keep this PR open and merge master back in rather than losing all this discussion on a closed PR.

**Comment by vnsavage** [comment]
Ok. Re-opened the PR - are you going to be working actively on it?

**Comment by joehoyle** [comment]
It's on my list - I'll probably hack on it next time I'm travelling!
